File: itcloud/src/test/java/com/itinfo/itcloud/service/storage/ItStorageServiceTest.java
package com.itinfo.itcloud.service.storage;

import com.itinfo.itcloud.model.IdentifiedVo;
import com.itinfo.itcloud.model.UsageVo;
import com.itinfo.itcloud.model.computing.ClusterVo;
import com.itinfo.itcloud.model.error.CommonVo;
import com.itinfo.itcloud.model.network.NetworkVo;
import com.itinfo.itcloud.model.storage.DiskVo;
import com.itinfo.itcloud.model.storage.DomainVo;
import com.itinfo.itcloud.model.storage.ImageCreateVo;
import org.apache.commons.fileupload.FileItem;
import org.apache.commons.fileupload.disk.DiskFileItem;
import org.apache.commons.io.IOUtils;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.multipart.commons.CommonsMultipartFile;

import java.io.*;
import java.nio.file.Files;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
class ItStorageServiceTest {
    @Autowired
    ItStorageService storageService;

    String dcId = "ae1d4138-f642-11ee-9c1b-00163e4b3128";


    @Test
    void totalStorage(){
        UsageVo result = storageService.totalStorage();
        System.out.println(result);
    }

    @Test
    @DisplayName("디스크 목록 출력")
    void getDiskList() {
        List<DiskVo> result = storageService.getDiskList(dcId);

        result.stream()
                .map(DiskVo::getAlias)
                .forEach(System.out::println);

        assertThat(storageService.getDiskList(dcId).size()).isEqualTo(38);
    }

    @Test
    @DisplayName("디스크 데이터센터 리스트")
    void setDcList(){
        List<IdentifiedVo> result = storageService.setDatacenterList();

        result.forEach(System.out::println);
    }

    @Test
    @DisplayName("디스크 스토리지 도메인 리스트")
    void setDomainList(){
        List<IdentifiedVo> result = storageService.setDomainList(dcId, null);

        result.forEach(System.out::println);
    }


    @Test
    @DisplayName("디스크 디스크 프로파일 리스트")
    void setDiskProfile(){
        String domainId = "12d17014-a612-4b6e-a512-6ec4e1aadba6";
//        String domainId = "43a786b3-e37e-4545-ba1f-5e3ae0ca6f0f";
        List<IdentifiedVo> result = storageService.setDiskProfile(domainId);

        result.forEach(System.out::println);
    }

    
    

    @Test
    @DisplayName("디스크 이미지 생성")
    void addDiskImage() {
        ImageCreateVo image =
                ImageCreateVo.builder()
                        .alias("S1")
                        .size(2)
                        .description("test")
                        .domainId("12d17014-a612-4b6e-a512-6ec4e1aadba6")
                        .profileId("23ab66ac-26c3-4b21-ba78-691ec2a004df")
                        .sparse(true)
                        .wipeAfterDelete(false)
                        .share(false)
                        .backup(true)
                        .build();

        CommonVo<Boolean> result = storageService.addDiskImage(image);
        assertThat(result.getHead().getCode()).isEqualTo(201);
//        assertThat("Sdd").isEqualTo(image.getAlias());
    }


    @Test
    @DisplayName("디스크 이미지 수정창")
    void setDiskImage(){
        String diskId = "d1947b58-84df-4ff3-802a-b4a6497853c6";
        ImageCreateVo result = storageService.setDiskImage(diskId);

        System.out.println(result);
    }


    @Test
    @DisplayName("디스크 이미지 수정")
    void editDiskImage() {
        String diskId = "d1947b58-84df-4ff3-802a-b4a6497853c6";

        ImageCreateVo image =
                ImageCreateVo.builder()
                        .id(diskId)
                        .alias("Sdd1")
                        .size(3)
                        .appendSize(0)
                        .description("tes2t")
                        .sparse(false)
                        .wipeAfterDelete(false)
                        .share(false)
                        .backup(true)
                        .build();

        CommonVo<Boolean> result = storageService.editDiskImage(image);

        assertThat(result.getHead().getCode()).isEqualTo(201);
    }


    @Test
    @DisplayName("디스크 이미지 삭제")
    void deleteDisk() {
        String diskId = "b0ded29b-04fc-451b-a7a4-bfd436f47890";
        CommonVo<Boolean> result = storageService.deleteDisk(diskId);

        assertThat(result.getHead().getCode()).isEqualTo(200);
    }

    @Test
    @DisplayName("디스크 이동창")
    void setDiskMoveDomainList() {
        String diskId = "f89493dd-51f8-44bd-9bfb-4687f43c822c";
        List<IdentifiedVo> result = storageService.setDomainList(dcId, diskId);

        result.forEach(System.out::println);
    }

    @Test
    @DisplayName("디스크 이동")
    void moveDisk() {
        String diskId = "f89493dd-51f8-44bd-9bfb-4687f43c822c";
        String domainId = "43a786b3-e37e-4545-ba1f-5e3ae0ca6f0f";
        CommonVo<Boolean> result = storageService.moveDisk(diskId, domainId);

        assertThat(result.getHead().getCode()).isEqualTo(200);
    }


    @Test
    @DisplayName("디스크 복사")
    void copyDisk() {
        String diskId = "f89493dd-51f8-44bd-9bfb-4687f43c822c";
        String domainId = "12d17014-a612-4b6e-a512-6ec4e1aadba6";
        DiskVo diskVo =
            DiskVo.builder()
                    .id(diskId)
                    .alias("abcdTest2")
                    .domainVo(DomainVo.builder().id(domainId).build())
                .build();

        CommonVo<Boolean> result = storageService.copyDisk(diskVo);

        assertThat(result.getHead().getCode()).isEqualTo(200);
    }


    @Test
    @DisplayName("디스크 이미지 업로드")
    void uploadDisk() throws IOException {
        // test환경에서는 실패할 경우 있음
        String path = "C:/Users/deh22/Documents/Rocky-8.4-x86_64-minimal.iso";

        File file = new File(path);
        FileItem fileItem = new DiskFileItem("file", Files.probeContentType(file.toPath()), false, file.getName(), (int) file.length(), file.getParentFile());
        InputStream inputStream = new FileInputStream(file);
        OutputStream outputStream = fileItem.getOutputStream();
        IOUtils.copy(inputStream, outputStream);
        MultipartFile multipartFile = new CommonsMultipartFile(fileItem);

        ImageCreateVo iVo =
                ImageCreateVo.builder()
                        .alias("absc")
                        .description("test")
                        .domainId("12d17014-a612-4b6e-a512-6ec4e1aadba6")
                        .profileId("23ab66ac-26c3-4b21-ba78-691ec2a004df")
                        .sparse(false)
                        .wipeAfterDelete(false)
                        .share(false)
                        .backup(false)
                        .build();

        CommonVo<Boolean> result = storageService.uploadDisk(multipartFile, iVo);

    }


    @Test
    @DisplayName("스토리지 도메인 목록")
    void getDomainList() {
        List<DomainVo> result = storageService.getDomainList(dcId);
        result.forEach(System.out::println);
    }

    @Test
    @DisplayName("호스트 목록")
    void setHostList() {
        List<IdentifiedVo> result = storageService.setHostList(dcId);

        result.forEach(System.out::println);
    }

    @Test
    void addDomain() {
    }

    @Test
    void deleteDomain() {
    }


    @Test
    @DisplayName("데이터센터 - 논리네트워크 목록")
    void getNetworkVoList() {
        List<NetworkVo> result = storageService.getNetworkVoList(dcId);

        result.forEach(System.out::println);
//        assertThat(8).isEqualTo(storageService.getNetworkVoList(dcId).size());
    }

    @Test
    @DisplayName("데이터센터 - 클러스터 목록")
    void getClusterVoList() {
        List<ClusterVo> result = storageService.getClusterVoList(dcId);

        result.forEach(System.out::println);
//        assertThat(1).isEqualTo(storageService.getClusterVoList(dcId).size());
    }

    @Test
    void getPermission() {
    }

    @Test
    void getEvent() {
    }


}

File: itcloud/src/test/java/com/itinfo/itcloud/service/storage/ItStorageServiceTest.kt
package com.itinfo.itcloud.service.storage

import com.itinfo.common.LoggerDelegate
import com.itinfo.itcloud.model.IdentifiedVo
import com.itinfo.itcloud.model.computing.DataCenterVo
import com.itinfo.itcloud.model.computing.EventVo
import com.itinfo.itcloud.model.computing.TemplateVo
import com.itinfo.itcloud.model.computing.VmVo
import com.itinfo.itcloud.model.setting.PermissionVo
import com.itinfo.itcloud.model.storage.DiskImageVo
import com.itinfo.itcloud.model.storage.DiskProfileVo
import com.itinfo.itcloud.model.storage.HostStorageVo
import com.itinfo.itcloud.model.storage.StorageDomainVo
import org.apache.commons.fileupload.FileItem
import org.apache.commons.fileupload.disk.DiskFileItem
import org.apache.commons.io.IOUtils

import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.multipart.commons.CommonsMultipartFile

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files

/**
 * [ItStorageServiceTest]
 * 스토리지 서비스 테스트
 *
 * @author chlee
 * @since 2024.07.15
 */
@SpringBootTest
class ItStorageServiceTest {
	@Autowired private lateinit var service: ItStorageService

	private lateinit var dataCenterId: String
	private lateinit var clusterId: String // Default
	private lateinit var networkId: String // ovirtmgmt(dc: Default)
	private lateinit var host01: String // host01
	private lateinit var domainId: String // hosted=engin
	private lateinit var nfs: String // hosted=engin

	@BeforeEach
	fun setup() {
		dataCenterId = "023b0a26-3819-11ef-8d02-00163e6c8feb"
		clusterId = "023c79d8-3819-11ef-bf08-00163e6c8feb"
		networkId = "00000000-0000-0000-0000-000000000009"
		host01 = "671e18b2-964d-4cc6-9645-08690c94d249"
		domainId = "571f10ef-3b23-417d-a954-d6ab90cf9ecb"
		nfs = "06faa572-f1ac-4874-adcc-9d26bb74a54d"
	}


	/**
	 * [should_findAll]
	 * [ItStorageService.findAll] 의 단위테스트
	 *
	 * @see [ItStorageService.findAll]
	 */
	@Test
	fun should_findAll() {
		log.debug("should_findAllDomains ... ")
		val result: List<StorageDomainVo> =
			service.findAll()

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
		assertThat(result.size, `is`(2))
	}

	/**
	 * [should_findAllFromDataCenter]
	 * [ItStorageService.findAllFromDataCenter] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllFromDataCenter]
	 */
	@Test
	fun should_findAllFromDataCenter() {
		log.debug("should_findAllFromDataCenter ... ")
		val result: List<StorageDomainVo> =
			service.findAllFromDataCenter(dataCenterId)

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
		assertThat(result.size, `is`(2))
	}

	/**
	 * [should_findOne]
	 * [ItStorageService.findOne] 의 단위테스트
	 *
	 * @see [ItStorageService.findOne]
	 */
	@Test
	fun should_findOne() {
		log.debug("should_findOne ... ")
		val result: StorageDomainVo? =
			service.findOne(domainId)

		assertThat(result, `is`(not(nullValue())))
		assertThat(result?.id, `is`(domainId))
		println(result)
	}

	/**
	 * [findAllIscsiFromHost]
	 * [ItStorageService.findAllIscsiFromHost] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllIscsiFromHost]
	 */
	@Test
	fun findAllIscsiFromHost() {
		log.debug("findAllIscsiFromHost ... ")
		val result: List<HostStorageVo> =
			service.findAllIscsiFromHost("7ba9acd9-90da-49f1-88f5-6ad82a246574")

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
		assertThat(result.size, `is`(0))
	}

	/**
	 * [findAllFibreFromHost]
	 * [ItStorageService.findAllFibreFromHost] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllFibreFromHost]
	 */
	@Test
	fun findAllFibreFromHost() {
		log.debug("findAllFibreFromHost ... ")
		val result: List<HostStorageVo> =
			service.findAllFibreFromHost("7ba9acd9-90da-49f1-88f5-6ad82a246574")

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
		assertThat(result.size, `is`(2))
	}



	/**
	 * [should_add]
	 * [ItStorageService.add] 의 단위테스트
	 *
	 * @see [ItStorageService.add]
	 */
	@Test
	fun should_add() {
		log.debug("should_add ... ")
	}

	/**
	 * [should_import]
	 * [ItStorageService.import] 의 단위테스트
	 *
	 * @see [ItStorageService.import]
	 */
	@Test
	fun should_import() {
		log.debug("should_import ... ")
	}

	/**
	 * [should_update]
	 * [ItStorageService.update] 의 단위테스트
	 *
	 * @see [ItStorageService.update]
	 */
	@Test
	fun should_update() {
		log.debug("should_update")
	}

	/**
	 * [should_remove]
	 * [ItStorageService.remove] 의 단위테스트
	 *
	 * @see [ItStorageService.remove]
	 */
	@Test
	fun should_remove() {
		log.debug("should_remove ... ")
	}

	/**
	 * [should_destroy]
	 * [ItStorageService.destroy] 의 단위테스트
	 *
	 * @see [ItStorageService.destroy]
	 */
	@Test
	fun should_destroy() {
		log.debug("should_destroy ... ")
	}


	/**
	 * [should_findAllDataCentersFromStorageDomain]
	 * [ItStorageService.findAllDataCentersFromStorageDomain] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllDataCentersFromStorageDomain]
	 */
	@Test
	fun should_findAllDataCentersFromStorageDomain() {
		log.debug("should_findAllDataCentersFromStorageDomain ... ")
		val result: List<DataCenterVo> =
			service.findAllDataCentersFromStorageDomain(domainId)

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
//		assertThat(result.size, `is`(1))
	}

	/**
	 * [should_attachFromDataCenter]
	 * [ItStorageService.attachFromDataCenter] 의 단위테스트
	 *
	 * @see [ItStorageService.attachFromDataCenter]
	 */
	@Test
	fun should_attachFromDataCenter() {
		log.debug("should_attachFromDataCenter ... ")
	}

	/**
	 * [should_detachFromDataCenter]
	 * [ItStorageService.detachFromDataCenter] 의 단위테스트
	 *
	 * @see [ItStorageService.detachFromDataCenter]
	 */
	@Test
	fun should_detachFromDataCenter() {
		log.debug("should_detachFromDataCenter ... ")
	}

	/**
	 * [should_activateFromDataCenter]
	 * [ItStorageService.activateFromDataCenter] 의 단위테스트
	 *
	 * @see [ItStorageService.activateFromDataCenter]
	 */
	@Test
	fun should_activateFromDataCenter() {
		log.debug("should_activateFromDataCenter ... ")
	}

	/**
	 * [should_maintenanceFromDataCenter]
	 * [ItStorageService.maintenanceFromDataCenter] 의 단위테스트
	 *
	 * @see [ItStorageService.maintenanceFromDataCenter]
	 */
	@Test
	fun should_maintenanceFromDataCenter() {
		log.debug("should_maintenanceFromDataCenter ... ")
	}



	/**
	 * [should_findAllVmsFromStorageDomain]
	 * [ItStorageService.findAllVmsFromStorageDomain] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllVmsFromStorageDomain]
	 */
	@Test
	fun should_findAllVmsFromStorageDomain() {
		log.debug("should_findAllVmsFromStorageDomain ... ")
		val result: List<VmVo> =
			service.findAllVmsFromStorageDomain(domainId)

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
		assertThat(result.size, `is`(3))
	}

	/**
	 * [should_findAllTemplatesFromStorageDomain]
	 * [ItStorageService.findAllTemplatesFromStorageDomain] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllTemplatesFromStorageDomain]
	 */
	@Test
	fun should_findAllTemplatesFromStorageDomain() {
		log.debug("should_findAllTemplatesFromStorageDomain ... ")
		val result: List<TemplateVo> =
			service.findAllTemplatesFromStorageDomain("072fbaa1-08f3-4a40-9f34-a5ca22dd1d74")

		assertThat(result, `is`(not(nullValue())))
		result.forEach { println(it) }
		assertThat(result.size, `is`(0))
	}


	/**
	 * [should_findAllEventsFromStorageDomain]
	 * [ItStorageService.findAllEventsFromStorageDomain] 의 단위테스트
	 *
	 * @see [ItStorageService.findAllEventsFromStorageDomain]
	 */
	@Test
	fun should_findAllEventsFromStorageDomain() {
		log.debug("should_findAllEventsFromStorageDomain ... ")
		val result: List<EventVo> =
			service.findAllEventsFromStorageDomain(domainId)

		assertThat(result, `is`(not(nullValue())))
		assertThat(result.size, `is`(33))
		println(result.size)
	}


	companion object {
		private val log by LoggerDelegate()
	}
}